c/dao.c

```diff
@@ -0,0 +1,290 @@
+#include "ckb_syscalls.h"
+#include "protocol_reader.h"
+
+#undef ns
+#define ns(x) FLATBUFFERS_WRAP_NAMESPACE(Ckb_Protocol, x)
+
+#define ERROR_UNKNOWN -1
+#define ERROR_WRONG_NUMBER_OF_ARGUMENTS -2
+#define ERROR_SYSCALL -4
+#define ERROR_BUFFER_NOT_ENOUGH -10
+#define ERROR_ENCODING -11
+#define ERROR_OVERFLOW -12
+#define ERROR_INVALID_WITHDRAW_BLOCK -13
+#define ERROR_INCORRECT_CAPACITY -14
+
+#define HASH_SIZE 32
+#define DAO_SIZE 32
+#define HEADER_SIZE 4096
+/* 32 KB */
+#define WITNESS_SIZE 32768
+
+#define LOCK_PERIOD_BLOCKS 10
+#define MATURITY_BLOCKS 5
+
+#define MAX(a, b) (((a) < (b)) ? (b) : (a))
+
+static int extract_bytes(ns(Bytes_table_t) bytes, unsigned char *buffer,
+                         volatile size_t *s) {
+  flatbuffers_uint8_vec_t seq = ns(Bytes_seq(bytes));
+  size_t len = flatbuffers_uint8_vec_len(seq);
+
+  if (len > *s) {
+    return ERROR_BUFFER_NOT_ENOUGH;
+  }
+
+  for (size_t i = 0; i < len; i++) {
+    buffer[i] = flatbuffers_uint8_vec_at(seq, i);
+  }
+  *s = len;
+
+  return CKB_SUCCESS;
+}
+
+/*
+ * Fetch withdraw header hash from the 3rd(offset by 1) argument
```

Comment:
From the code, it seems that you are getting the 2nd argument.

Reply:
Oops this is due to recoverable signature now, I forgot to update the comment here.

Reply:
❓Does it limit what kind of lock scripts can be used when deposit into DAO?

- The lock script must use the first witness to unlock
- The lock script must work without any knowledge of the following witness arguments.
- The lock script must protect the following witness arguments against arbitrary modification.

Reply:
Yes it limits a lock script in some way, but I do want to point out that a lock script should prevent the non-signature arguments from tampering anyway, otherwise we would have a tx malleability problem